// @walletconnect/utils/dist/cjs/session.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.parseWalletConnectUri = exports.isWalletConnectSession = void 0;

var url_1 = require("./url");

function isWalletConnectSession(object) {
    return typeof object.bridge !== "undefined";
}

exports.isWalletConnectSession = isWalletConnectSession;

function parseWalletConnectUri(str) {
    var pathStart = str.indexOf(":");
    var pathEnd = str.indexOf("?") !== -1 ? str.indexOf("?") : undefined;
    var protocol = str.substring(0, pathStart);
    var path = str.substring(pathStart + 1, pathEnd);

    function parseRequiredParams(path) {
        var separator = "@";
        var values = path.split(separator);
        var requiredParams = {
            handshakeTopic: values[0],
            version: parseInt(values[1], 10)
        };
        return requiredParams;
    }

    var requiredParams = parseRequiredParams(path);
    var queryString = typeof pathEnd !== "undefined" ? str.substr(pathEnd) : "";

    function parseQueryParams(queryString) {
        var result = url_1.parseQueryString(queryString);
        var parameters = {
            key: result.key || "",
            bridge: result.bridge || ""
        };
        return parameters;
    }

    var queryParams = parseQueryParams(queryString);
    var result = Object.assign(Object.assign({
        protocol: protocol
    }, requiredParams), queryParams);
    return result;
}

exports.parseWalletConnectUri = parseWalletConnectUri;
