// multibase/src/base.js
'use strict';

var _classCallCheck = require("/Applications/XAMPP/xamppfiles/htdocs/aa-superlion-mainnet/node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/classCallCheck");

var _createClass = require("/Applications/XAMPP/xamppfiles/htdocs/aa-superlion-mainnet/node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/createClass");

var Base = /*#__PURE__*/ function() {
    function Base(name, code, implementation, alphabet) {
        _classCallCheck(this, Base);

        this.name = name;
        this.code = code;
        this.alphabet = alphabet;

        if (implementation && alphabet) {
            this.engine = implementation(alphabet);
        }
    }

    _createClass(Base, [{
        key: "encode",
        value: function encode(stringOrBuffer) {
            return this.engine.encode(stringOrBuffer);
        }
    }, {
        key: "decode",
        value: function decode(stringOrBuffer) {
            return this.engine.decode(stringOrBuffer);
        }
    }, {
        key: "isImplemented",
        value: function isImplemented() {
            return this.engine;
        }
    }]);

    return Base;
}();

module.exports = Base;
