// fs-extra/lib/move-sync/move-sync.js
'use strict';

var fs = require('graceful-fs');

var path = require('path');

var copySync = require('../copy-sync').copySync;

var removeSync = require('../remove').removeSync;

var mkdirpSync = require('../mkdirs').mkdirpSync;

var stat = require('../util/stat');

function moveSync(src, dest, opts) {
    opts = opts || {};
    var overwrite = opts.overwrite || opts.clobber || false;

    var _stat$checkPathsSync = stat.checkPathsSync(src, dest, 'move'),
        srcStat = _stat$checkPathsSync.srcStat;

    stat.checkParentPathsSync(src, srcStat, dest, 'move');
    mkdirpSync(path.dirname(dest));
    return doRename(src, dest, overwrite);
}

function doRename(src, dest, overwrite) {
    if (overwrite) {
        removeSync(dest);
        return rename(src, dest, overwrite);
    }

    if (fs.existsSync(dest)) throw new Error('dest already exists.');
    return rename(src, dest, overwrite);
}

function rename(src, dest, overwrite) {
    try {
        fs.renameSync(src, dest);
    } catch (err) {
        if (err.code !== 'EXDEV') throw err;
        return moveAcrossDevice(src, dest, overwrite);
    }
}

function moveAcrossDevice(src, dest, overwrite) {
    var opts = {
        overwrite: overwrite,
        errorOnExist: true
    };
    copySync(src, dest, opts);
    return removeSync(src);
}

module.exports = moveSync;
