// @walletconnect/browser-utils/dist/cjs/browser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getClientMeta = exports.getLocalStorage = exports.getLocalStorageOrThrow = exports.getCrypto = exports.getCryptoOrThrow = exports.getLocation = exports.getLocationOrThrow = exports.getNavigator = exports.getNavigatorOrThrow = exports.getDocument = exports.getDocumentOrThrow = exports.getFromWindowOrThrow = exports.getFromWindow = exports.isBrowser = exports.isNode = exports.isMobile = exports.isIOS = exports.detectOS = exports.detectEnv = void 0;

var tslib_1 = require("tslib");

var windowMetadata = tslib_1.__importStar(require("window-metadata"));

var windowGetters = tslib_1.__importStar(require("window-getters"));

var detect_browser_1 = require("detect-browser");

function detectEnv(userAgent) {
    return detect_browser_1.detect(userAgent);
}

exports.detectEnv = detectEnv;

function detectOS() {
    var env = detectEnv();
    return env && env.os ? env.os : undefined;
}

exports.detectOS = detectOS;

function isIOS() {
    var os = detectOS();
    return os ? os.toLowerCase().includes("ios") : false;
}

exports.isIOS = isIOS;

function isMobile() {
    var os = detectOS();
    return os ? os.toLowerCase().includes("android") || os.toLowerCase().includes("ios") : false;
}

exports.isMobile = isMobile;

function isNode() {
    var env = detectEnv();
    var result = env && env.name ? env.name.toLowerCase() === "node" : false;
    return result;
}

exports.isNode = isNode;

function isBrowser() {
    var result = !isNode() && !!exports.getNavigator();
    return result;
}

exports.isBrowser = isBrowser;
exports.getFromWindow = windowGetters.getFromWindow;
exports.getFromWindowOrThrow = windowGetters.getFromWindowOrThrow;
exports.getDocumentOrThrow = windowGetters.getDocumentOrThrow;
exports.getDocument = windowGetters.getDocument;
exports.getNavigatorOrThrow = windowGetters.getNavigatorOrThrow;
exports.getNavigator = windowGetters.getNavigator;
exports.getLocationOrThrow = windowGetters.getLocationOrThrow;
exports.getLocation = windowGetters.getLocation;
exports.getCryptoOrThrow = windowGetters.getCryptoOrThrow;
exports.getCrypto = windowGetters.getCrypto;
exports.getLocalStorageOrThrow = windowGetters.getLocalStorageOrThrow;
exports.getLocalStorage = windowGetters.getLocalStorage;

function getClientMeta() {
    return windowMetadata.getWindowMetadata();
}

exports.getClientMeta = getClientMeta;
