// @walletconnect/browser-utils/dist/cjs/mobile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getMobileLinkRegistry = exports.getMobileRegistryEntry = exports.saveMobileLinkInfo = exports.formatIOSMobile = exports.mobileLinkChoiceKey = void 0;

var local_1 = require("./local");

exports.mobileLinkChoiceKey = "WALLETCONNECT_DEEPLINK_CHOICE";

function formatIOSMobile(uri, entry) {
    var encodedUri = encodeURIComponent(uri);
    return entry.universalLink ? "".concat(entry.universalLink, "/wc?uri=").concat(encodedUri) : entry.deepLink ? "".concat(entry.deepLink).concat(entry.deepLink.endsWith(":") ? "//" : "/", "wc?uri=").concat(encodedUri) : "";
}

exports.formatIOSMobile = formatIOSMobile;

function saveMobileLinkInfo(data) {
    var focusUri = data.href.split("?")[0];
    local_1.setLocal(exports.mobileLinkChoiceKey, Object.assign(Object.assign({}, data), {
        href: focusUri
    }));
}

exports.saveMobileLinkInfo = saveMobileLinkInfo;

function getMobileRegistryEntry(registry, name) {
    return registry.filter(function(entry) {
        return entry.name.toLowerCase().includes(name.toLowerCase());
    })[0];
}

exports.getMobileRegistryEntry = getMobileRegistryEntry;

function getMobileLinkRegistry(registry, whitelist) {
    var links = registry;

    if (whitelist && whitelist.length) {
        links = whitelist.map(function(name) {
            return getMobileRegistryEntry(registry, name);
        });
    }

    return links;
}

exports.getMobileLinkRegistry = getMobileLinkRegistry;
